TST Interaction between `class_weight` and `sample_weight`

vitaliset opened the pull request:
**Reference Issues/PRs**

Towards #21504.

**What does this implement/fix? Explain your changes.**

This pull request introduces two test functions for scikit-learn classifier that check the interaction between `class_weight` and `sample_weight` parameters as proposed by @jjerphan:

- `check_interaction_of_class_and_sample_weight_excluding_class`: Verifies that setting a class weight to zero is equivalent to excluding samples of that class from calibration, even with non-uniform sample weights.

- `check_interaction_of_class_and_sample_weight_excluding_samples`: Confirms that setting sample weights to zero is equivalent to removing corresponding samples from training, even with non-uniform class weights.

**Any other comments?**

- Classifiers that failed for a reason other than `assert_allclose_dense_sparse`:.

    - `NuSVC` failed the `check_sample_weights_invariance` test (present check) due to an error in the value of the nu parameter. This error persisted here.

    - `SVC` is giving an error due to "dual coefficients or intercepts are not finite".

    - `LinearSVC` is failing in the `assert_allclose_dense_sparse`, but not for the reasons one would expect. The output of `decision_function` method is giving an array of shape `(n_samples,)`, instead of `(n_samples, n_classes)` when trained with zero sample weight for one of the classes (and `n_classes=3`). I suppose this is a bug. (Worthy of an issue?)

<details>
  <summary>Minimal code with the same error of SVC</summary>

```python
from sklearn import __version__
print(__version__)
>>> 1.2.0
from sklearn.svm import SVC
from sklearn.datasets import make_classification

n_classes = 3
X, y = make_classification(
    n_samples=200,
    n_classes=n_classes,
    n_informative=2 * n_classes,
    random_state=0,
)
class_weight_dict = {cls: 1 if cls != 0 else 0 for cls in range(n_classes)}
print(class_weight_dict)
>>> {0: 0, 1: 1, 2: 1}

svc = SVC(class_weight=class_weight_dict).fit(X, y)
>>> ValueError: The dual coefficients or intercepts are not finite. The input data may contain large values and need to bepreprocessed.
```
</details>


<details>
  <summary>Minimal code with the same error of LinearSVC</summary>

```python
from sklearn import __version__
print(__version__)
>>> 1.2.0
import numpy as np
from sklearn.datasets import make_classification
from sklearn.svm import LinearSVC

n_classes = 3
X, y = make_classification(
    n_samples=200,
    n_classes=n_classes,
    n_informative=2 * n_classes,
    random_state=0,
)

sample_weight = np.ones_like(y)
sample_weight_zero_weight_first = np.where(y == 0, 0, sample_weight)

svc_sw = (
    LinearSVC(random_state=0)
    .fit(X, y, sample_weight=sample_weight_zero_weight_first)
)
>>> ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.

print(svc_sw.decision_function(X).shape)
>>> (200,)

svc = (
    LinearSVC(random_state=0)
    .fit(X, y, sample_weight=sample_weight)
)
>>> ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.

print(svc.decision_function(X).shape)
>>> (200, 3)
```
</details>

vitaliset committed:
draft commit

vitaliset committed:
black linting and small adjustments

vitaliset committed:
more linting

jjerphan reviewed (commented):
Hi @vitaliset, thank you for this PR.

FYI, scikit-learn's maintainers prefer using [Draft PR](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/changing-the-stage-of-a-pull-request) instead of prefixing their title with `[WIP]` or such. I have changed the state of this branch accordingly to those conventions.

I do not have time to review this PR for now, but I think if tests are failing due to faulty logic in scikit-learn, then the best is to address those problems separately using several PRS.

What do you think of this?

vitaliset commented:
I hadn't realized that I could set the draft flag myself! Thank you so much for the information, @jjerphan!

Currently, I am performing checks on all estimators using`_yield_classifier_checks` within the `check_estimator`. Should I create an ignore list that my function would utilize to bypass certain estimators? Or should I implement the logic within distinct tests tailored for each estimator? I'm not a big fan of the second approach because I would be replicating the code in multiple places as far as I see.

jjerphan commented:
For now, I think we can start with a `xfail`-list. Then we could create a meta-issue to empty this list by fixing individual problems.

vitaliset committed:
add if for having sample_weight in fit

vitaliset committed:
Merge branch 'main' into tst_interaction_class_weight_sample_weight

vitaliset committed:
Merge branch 'main' into tst_interaction_class_weight_sample_weight

vitaliset committed:
add xfail for LinearSVC, NuSVC, Perceptron and SGDClassifier

vitaliset committed:
add if not pairwise to skip precomputed and add binary_only check to see max_classes

vitaliset committed:
correct interaction for checking binary_only

vitaliset committed:
small comments and add xfail to SVC

vitaliset commented:
Hello @jjerphan, I have implemented the xfail-logic to the estimators failing due to faulty coherence in them, resulting in a green CI. I have update the PR description to make this step explicit. I believe the PR is now in a suitable state for a review so I'll remove the draft flag.

I understand you might be short on time at the moment, but I wanted to let you know on this update. :)

jjerphan commented on sklearn/utils/estimator_checks.py:
```diff
@@ -3050,6 +3055,120 @@ def check_class_weight_balanced_linear_classifier(name, Classifier):
+        for method in ["predict", "predict_proba", "decision_function", "transform"]:
+            if hasattr(estimator_orig, method):
+                pred_sw = getattr(estimator_sw, method)(X)
+                pred_exclude = getattr(estimator_sw, method)(X)
```
```suggestion
                pred_exclude = getattr(estimator_exclude, method)(X)
```

vitaliset replied:
Let's pretend this didn't happen 🫠 hahahah

jjerphan commented on sklearn/utils/estimator_checks.py:
```diff
@@ -3050,6 +3055,120 @@ def check_class_weight_balanced_linear_classifier(name, Classifier):
+            if hasattr(estimator_orig, method):
+                pred_cw = getattr(estimator_cw, method)(X)
+                pred_sw = getattr(estimator_sw, method)(X)
+                pred_exclude = getattr(estimator_sw, method)(X)
```
```suggestion
                pred_exclude = getattr(estimator_exclude, method)(X)
```

vitaliset committed:
Apply suggestions from code review

Co-authored-by: Julien Jerphanion <git@jjerphan.xyz>

vitaliset committed:
remove all xfail to see ci results

vitaliset committed:
add transformation for correct shape

vitaliset committed:
incorrect transformation for 2 classes

vitaliset committed:
second test does not need shape transformation

vitaliset committed:
run ci without xfail with jjerphans correction

vitaliset committed:
correcting test names

vitaliset committed:
linting

vitaliset committed:
removing extra debugging message

vitaliset committed:
adding all the xfails again

vitaliset committed:
missing one xfail

vitaliset committed:
extra xfail

jjerphan commented on sklearn/utils/estimator_checks.py:
```diff
@@ -3105,12 +3105,18 @@ def check_interaction_of_class_and_sample_weight_excluding_class(name, estimator
 
         # Checking if the output is the same for multiple outputs
-        for method in ["predict", "predict_proba", "decision_function", "transform"]:
+        for method in ["predict_proba", "decision_function", "predict"]:
```
Why has `tranform` been removed here?

vitaliset replied:
I removed it at first because we are dealing only with classifiers, but it has been added back in on the latest commit! :)

By the way, not a problem if you can't review it for now. If you have any other core developers in mind who could take a look, I can work on fixing any issues they identify in the meantime. Alternatively, I can reach out to you again in a couple of months to continue our work from where we left off.

jjerphan commented on sklearn/utils/estimator_checks.py:
```diff
@@ -3162,10 +3168,10 @@ def check_interaction_of_class_and_sample_weight_excluding_samples(
 
         # Checking if the output is the same for multiple outputs
-        for method in ["predict", "predict_proba", "decision_function", "transform"]:
+        for method in ["predict_proba", "decision_function", "predict"]:
```
Ditto: why has `tranform` been removed here?

jjerphan reviewed (commented):
A few more comments.

Unfortunately, I have no bandwidth to properly review this now.

vitaliset committed:
adding transform back

vitaliset committed:
Merge branch 'main' into tst_interaction_class_weight_sample_weight

jjerphan commented on sklearn/ensemble/_forest.py:
```diff
@@ -1442,6 +1442,20 @@ def __init__(
+        return {
+            "_xfail_checks": {
+                "check_interaction_of_class_and_sample_weight_excluding_samples": (
+                    "while using class weight, setting some sample's weight to zero "
```
This suggestion also applies in many other places.
```suggestion
                    "while using class weight, setting some samples' weights to zero "
```

jjerphan commented on sklearn/utils/estimator_checks.py:
```diff
@@ -3050,6 +3055,126 @@ def check_class_weight_balanced_linear_classifier(name, Classifier):
+        "equivalent to excluding the samples from that class."
+    )
+
+    for n_classes in range(3, 10):
```
Can you parametrize this test on `n_classes` using `pytest.mark.parametrize`?

vitaliset replied:
I'm afraid I don't know how to do it. I have added the decorator to the check functions, but they don't "propagate" to the test itself. These check functions are being yield by `_yield_all_checks` inside `parametrize_with_checks` where there is another `pytest.mark.parametrize` that must mess up with the previous one. I don't know...

```python
@pytest.mark.parametrize("n_classes", range(3, 10))
def check_interaction_of_class_and_sample_weight_excluding_class(
    name, estimator_orig, n_classes
):
...
```

Output from tests:
```python
 TypeError: check_interaction_of_class_and_sample_weight_excluding_class() missing 1 required positional argument: 'n_classes'
```

Is there another way to do what you asked?

jjerphan commented on sklearn/utils/estimator_checks.py:
```diff
@@ -3050,6 +3055,126 @@ def check_class_weight_balanced_linear_classifier(name, Classifier):
+
+                pred_exclude = getattr(estimator_exclude, method)(X)
+                if method in ["predict_proba", "decision_function"]:
+                    pred_cw = pred_cw[:, list(range(1, n_classes))]
```
```suggestion
                    # Excluding the first class (class 0)
                    pred_cw = pred_cw[:, 1:]
```

jjerphan commented on sklearn/utils/estimator_checks.py:
```diff
@@ -3050,6 +3055,126 @@ def check_class_weight_balanced_linear_classifier(name, Classifier):
+                pred_exclude = getattr(estimator_exclude, method)(X)
+                if method in ["predict_proba", "decision_function"]:
+                    pred_cw = pred_cw[:, list(range(1, n_classes))]
+                    if method == "decision_function" and n_classes == 3:
+                        pred_cw = pred_cw[:, 1]
```
Can you explain this, please?

vitaliset replied:
Because we excluded one class, for `n_classes` equal to 3, we are in a binary classification setting. The decision function outputs only one class output (similar to how you can get `predict_proba` of one class as the complement of another). Because of this, we have to exclude one column of the 3-class output to compare the arrays.

If I leave it as is I get the error:
```python
E           AssertionError:
...
E           (shapes (200, 2), (200,) mismatch)
E            x: array([[ 0.103091, -0.103091],
E                  [ 0.582987, -0.582987],
E                  [ 1.004391, -1.004391],...
E            y: array([-0.103091, -0.582987, -1.004391,  0.340705,  0.262283,  0.570671,
E                  -0.042541, -0.337614, -0.307904, -0.036075, -0.441613,  0.261148,
E                  -0.196478, -0.808903, -0.903717,  0.097252, -1.368786, -0.059082,...
```

I'm adding a comment to explain this.

jjerphan commented on sklearn/utils/estimator_checks.py:
```diff
@@ -3050,6 +3055,126 @@ def check_class_weight_balanced_linear_classifier(name, Classifier):
+    tags = _safe_tags(estimator_orig)
+    max_classes = 2 if tags["binary_only"] else 9
+
+    for n_classes in range(2, max_classes + 1):
```
Can you parametrize this test on `n_classes` using `pytest.mark.parametrize`?

jjerphan commented on sklearn/utils/estimator_checks.py:
```diff
@@ -3050,6 +3055,126 @@ def check_class_weight_balanced_linear_classifier(name, Classifier):
+        estimator_sw.fit(X, y, sample_weight=sample_weight_zero_weight_first)
+
+        # Checking if the output is the same for multiple outputs
+        for method in ["predict", "predict_proba", "decision_function", "transform"]:
```
Could you define `err_msg_*` here to include `method`? This way inspecting cases which might fail will be easier.

Alternatively, we could parameterize this test of `method` if the test execution is negligible.

vitaliset replied:
Done it. Not resolving the comment just yet because I may want to parameterize it if we manage to do it.

jjerphan commented on sklearn/utils/estimator_checks.py:
```diff
@@ -3050,6 +3055,126 @@ def check_class_weight_balanced_linear_classifier(name, Classifier):
+        set_random_state(estimator_exclude, random_state=0)
+        estimator_exclude.fit(X_exclude, y_exclude)
+
+        # Checking if the output is the same for multiple outputs
+        for method in ["predict", "predict_proba", "decision_function", "transform"]:
+            if hasattr(estimator_orig, method):
+                pred_sw = getattr(estimator_sw, method)(X)
+                pred_exclude = getattr(estimator_exclude, method)(X)
+                assert_allclose_dense_sparse(pred_sw, pred_exclude, err_msg=err_msg)
```
The previous comment also applies here for `err_msg`.

jjerphan commented on sklearn/utils/estimator_checks.py:
```diff
@@ -164,6 +165,10 @@ def _yield_classifier_checks(classifier):
         yield check_estimators_unfitted
     if "class_weight" in classifier.get_params().keys():
         yield check_class_weight_classifiers
+        if has_fit_parameter(classifier, "sample_weight") and not tags["pairwise"]:
```
This is not obvious why the test operates on `tags["pairwise"]`.

Can you add a comment precising it?

vitaliset replied:
I added it on aab209a trying to solve some xfail logic, but I removed it on this last commit and I did not get any error. I think it is not needed (?).

jjerphan commented on sklearn/utils/estimator_checks.py:
```diff
@@ -3050,6 +3055,126 @@ def check_class_weight_balanced_linear_classifier(name, Classifier):
+        estimator_exclude.fit(X_exclude, y_exclude, sample_weight=sample_weight_exclude)
+
+        # Model setting `sample_weight` to 0 for samples from class 0
+        sample_weight_zero_weight_first = np.where(y == 0, 0, sample_weight)
```
I ticked on this name, but I am unable to find any other better one. Let's use this one!

jjerphan reviewed (commented):
I got some time between two rays of sunny weather!

LGTM modulo a few last comments.

EwoutH commented:
Looks useful to make to tests these interactions properly. Thanks for the effort! What's needed to move this PR forward?

vitaliset commented:
Coincidentally, I restarted working on another old contribution of mine to scikit-learn last week and I intend to finish the work on this one next!

If I don't have time for that within the next few weeks, I'll let you know @EwoutH and you can pick up from where I left off (if you want to).

vitaliset committed:
Merge branch 'main' into tst_interaction_class_weight_sample_weight

vitaliset committed:
Apply suggestions from code review
Co-authored-by: Julien Jerphanion <git@jjerphan.xyz>

vitaliset committed:
better err_msg, excluding the tags["pairwise"] check, comment explaining decision_function n_classes 3

vitaliset commented:
Hello @jjerphan! I corrected and updated the PR with your review, except for an issue when parameterizing the test. I left some comments on your review so we can discuss what's missing or resolve them.

Sorry for the delay in responding to this review. I haven't worked on scikit-learn since mid-last year, but I'm excited to start contributing again! :)
